src/System.IO.Pipelines/Pipe.cs

```diff
@@ -472,13 +481,19 @@ void IPipeReader.Complete(Exception exception)
                 PipelinesThrowHelper.ThrowInvalidOperationException(ExceptionResource.CompleteReaderActiveReader, _readingState.Location);
             }
 
-            _readerCompletion.TryComplete(exception);
-
+            bool runCompletionCallbacks;
             Action awaitable;
+
             lock (_sync)
             {
+                runCompletionCallbacks = _readerCompletion.TryComplete(exception);
                 awaitable = _writerAwaitable.Complete();
             }
+
+            if (runCompletionCallbacks)
+            {
+                TrySchedule(_writerScheduler, _callReaderCompletedCallbacks, this);
+            }
             TrySchedule(_writerScheduler, awaitable);
 
             if (_writerCompletion.IsCompleted)
```

Comment:
I think these do need to be inside of the lock. What happens if `_writerCompletion.IsCompleted` becomes true after this check? Won't it just leak memory?